### DISCUSSION TITLE
Unexpected issue regarding `addon-fit` when used in React

### DISCUSSION THREAD

**Discussion by brandonmcconnell** [sourcelair/xterm.js#5017] [Q&A]
I have a minimal reproduction here, where you can see some issues in the console and client that appear, apparently related to the usage of `addon-fit`.
https://stackblitz.com/edit/stackblitz-starters-6ssv2a?file=app%2Fcomponents%2FXTerm.tsx

Do you have any idea what I could be doing wrong in my wrapper component? It's a very thin wrapper, so I feel the issue is probably related to my usage of XTerm and not React-specific.

The component source and errors are below:

### Component source

<details open><summary><h4>app/page.tsx</h4></summary>

```tsx
import XTerm from './components/XTerm';

export default function Home() {
  return (
    <main className="flex min-h-screen flex-col items-center justify-between p-24">
      <XTerm />
    </main>
  );
}
```

</details>

<details open><summary><h4>app/components/XTerm.tsx</h4></summary>

```tsx
'use client';

import React, { useEffect, useRef } from 'react';
import '@xterm/xterm/css/xterm.css';

// Import addons

// https://github.com/xtermjs/xterm.js
import { Terminal } from '@xterm/xterm';
// https://github.com/xtermjs/xterm.js/tree/master/addons/addon-fit
import { FitAddon } from '@xterm/addon-fit';
// https://github.com/xtermjs/xterm.js/tree/master/addons/addon-web-links
import { SearchAddon } from '@xterm/addon-search';
// https://github.com/xtermjs/xterm.js/tree/master/addons/addon-image
import { WebLinksAddon } from '@xterm/addon-web-links';
// https://github.com/xtermjs/xterm.js/tree/master/addons/addon-search
import { ImageAddon } from '@xterm/addon-image';

const XTerm = () => {
  const terminalRef = useRef(null);
  const terminal = useRef(new Terminal());

  useEffect(() => {
    // ComponentDidMount
    if (terminalRef.current) {
      const fitAddon = new FitAddon();
      const searchAddon = new SearchAddon();
      const webLinksAddon = new WebLinksAddon();
      const imageAddon = new ImageAddon({ pixelLimit: 16777216 });

      // Set up addons
      terminal.current.loadAddon(fitAddon);
      terminal.current.loadAddon(searchAddon);
      terminal.current.loadAddon(webLinksAddon);
      terminal.current.loadAddon(imageAddon);

      // Attach the terminal to the div
      terminal.current.open(terminalRef.current);
      fitAddon.fit();
      // Example output
      terminal.current.writeln('Hello world');
    }

    // ComponentWillUnmount
    return () => terminal.current.dispose(); // Cleanup

    // Empty dependency array, to only run on mount and cleanup on unmount
  }, []);

  return <div ref={terminalRef} style={{ height: '300px', width: '600px' }} />;
};

export default XTerm;
```

</details>

### Errors

In the terminal:
```
 X node_modules/@xterm/addon-fit/lib/addon-fit.js (1:196) @ eval
 X ReferenceError: self is not defined
    at __webpack_require__ (/home/projects/stackblitz-starters-6ssv2a/.next/server/webpack-runtime.js:33:42)
    at eval (./app/components/XTerm.tsx:12:74)
    at (ssr)/./app/components/XTerm.tsx (/home/projects/stackblitz-starters-6ssv2a/.next/server/app/page.js:151:1)
    at __webpack_require__ (/home/projects/stackblitz-starters-6ssv2a/.next/server/webpack-runtime.js:33:42)
```

In the client:
```
Unhandled Runtime Error
TypeError: Cannot read properties of undefined (reading 'dimensions')

Call Stack
get dimensions
node_modules/@xterm/xterm/lib/xterm.js (1:108798)
t.Viewport.syncScrollArea
node_modules/@xterm/xterm/lib/xterm.js (1:50747)
eval
node_modules/@xterm/xterm/lib/xterm.js (1:48246)
```

**Comment by brandonmcconnell** [comment]
Per [this answer](https://github.com/xtermjs/xterm.js/discussions/4895#discussioncomment-8138759) by @x1unix on [this thread](https://github.com/xtermjs/xterm.js/discussions/4895), I also tried doing this in a fully client setting (no Next.js) and used [the source](https://github.com/jaxtonw/xterm-for-react18/blob/master/src/XTerm.tsx) from https://github.com/jaxtonw/xterm-for-react18 as a baseline, and it still does not work.

**Reproduction:** https://stackblitz.com/edit/vitejs-vite-mpuuoi?file=src%2Fcomponents%2FXTerm.tsx

**Comment by JeanMeche** [comment]
We have the same issue over for angular.dev that uses pre-rendering. `self` only exist in browser environment and prerendering runs on node.

**Reply by jerch** [reply]
Pre-rendering as in server-side-rendering? Thats not a good idea with xterm.js and will most likely fail miserably - xterm.js uses tons of browser states for IO and screen handling. Unless your prerendering environment almost fully emulates a browser engine... (I am not aware of any)

**Reply by JeanMeche** [reply]
The import has side effects, so even if xterm only runs in the browser, the bare import broke the pre-rendering part.

**Reply by jerch** [reply]
As I wrote - it needs an almost full browser environment. If the import already breaks, then your pre-rendering env is far from "almost full".

**Reply by JeanMeche** [reply]
Well #4940 fixed the issue for me. Which should also fix the issue mentionned by OP.
